### PR TITLE
[Automation] Slack Alerts

### PR DESCRIPTION
**Pull request by bh2smith**
Based on the discussion in slack: https://cowservices.slack.com/archives/C0361CDD1FZ/p1658843635716449

As the final step before automation of this process, we needed a way to alert the team that a transaction has been created. This PR introduces basic functionality required to alert on the posted transaction in the following way:

Once the transaction is posted to the safe, a message is sent to our internal #dev-multisig channel

```
Solver Rewards transaction with nonce {nonce} pending signatures.
To sign and execute, visit:
{SAFE_URL}
More details in thread
```

Inside the thread of this message, we also post the relevant logs. Specifically,

<details>

```sh
The data being aggregated here is available for visualization at
https://dune.com/gnosis.protocol/solver-rewards-accounting?StartTime=2022-07-10+00%3A00%3A00&EndTime=2022-07-17+00%3A00%3A00
Deducting slippage for solver 0x109BF9E0287Cc95cc623FBE7380dD841d4bdEb03by -0.01976 (barn-Otex)
Deducting slippage for solver 0x22DEE0935c77d32c7241362b14e76fc2D5eF657dby -0.00000 (barn-BalancerSOR)
Deducting slippage for solver 0x3Cee8C7d9B5C8F225A8c36E7d3514e1860309651by -0.00157 (prod-Baseline)
Deducting slippage for solver 0x5B0BFe439ab45A4F002C259b1654ED21c9a42D69by -0.01116 (barn-PLM)
Deducting slippage for solver 0x7A0A8890D71A4834285efDc1D18bb3828e765C6Aby -0.00067 (prod-Naive)
Deducting slippage for solver 0x8567351D6989d83513D3BC3ad951CcCe363941e3by -0.00175 (barn-Atlas)
Deducting slippage for solver 0xa0044c620Da7f2876dA7004719b8370eb7BE5e50by -0.00431 (barn-MIP)
Deducting slippage for solver 0xb20B86C4e6DEEB432A22D773a221898bBBD03036by -4.44768 (prod-1inch)
Deducting slippage for solver 0xc9ec550BEA1C64D779124b23A26292cc223327b6by -0.51984 (prod-Otex)
Deducting slippage for solver 0xda324c2f06d3544E7965767ce9ca536dcb67a660by -0.01314 (barn-QuasiModo)
Deducting slippage for solver 0xDe786877a10DBb7EBa25a4DA65aEcf47654F08abby -0.00683 (barn-0x)
Deducting slippage for solver 0xe18B5632DF2Ec339228DD65e4D9F004eF59653d3by -0.50232 (prod-Atlas)
Deducting slippage for solver 0xe33062a24149f7801a48B2675Ed5111d3278F0f5by -0.05856 (barn-1inch)
Slippage for 0xe33062a24149f7801a48B2675Ed5111d3278F0f5(barn-1inch) exceeds reimbursement: Invalid adjustment TransferETH(receiver=0xe33062a24149f7801a48B2675Ed5111d3278F0f5, amount_wei=0.0409522703268333) by -0.05856302842290827
Excluding payout and appending excess to overdraft
Deducting slippage for solver 0xE9aE2D792F981C53EA7f6493A17Abf5B2a45a86bby -1.94828 (prod-ParaSwap)
Deducting slippage for solver 0xF7995B6B051166eA52218c37b8d03A2A6bbeF3DAby -0.26270 (prod-BalancerSOR)
Redirecting solver 0x080a8b1E2f3695E179453c5e617b72A381BE44B9 COW tokens (480.0) to 0x6C642caFCbd9d8383250bb25F67aE409147f78b2
Redirecting solver 0x0d2584DA2F637805071f184bCFa1268eBAe8a24A COW tokens (170.0) to 0x6C642caFCbd9d8383250bb25F67aE409147f78b2
Redirecting solver 0x109BF9E0287Cc95cc623FBE7380dD841d4bdEb03 COW tokens (3570.0) to 0x84dbAE2549d67CaF00f65C355de3D6F4df59A32C
Redirecting solver 0x149d0f9282333681Ee41D30589824b2798E9fb47 COW tokens (82355.0) to 0xf7f9C5ecAFd159271fe55A8FEdEDa95d7fcD0DCc
Redirecting solver 0x22DEE0935c77d32c7241362b14e76fc2D5eF657d COW tokens (170.0) to 0x6C642caFCbd9d8383250bb25F67aE409147f78b2
Redirecting solver 0x3Cee8C7d9B5C8F225A8c36E7d3514e1860309651 COW tokens (7970.0) to 0x6C642caFCbd9d8383250bb25F67aE409147f78b2
Redirecting solver 0x5B0BFe439ab45A4F002C259b1654ED21c9a42D69 COW tokens (510.0) to 0xf7f9C5ecAFd159271fe55A8FEdEDa95d7fcD0DCc
Redirecting solver 0x6D1247b8ACf4dFD5Ff8cfD6C47077ddC43d4500E COW tokens (7850.0) to 0x6C642caFCbd9d8383250bb25F67aE409147f78b2
Redirecting solver 0x731a0A8ab2C6FcaD841e82D06668Af7f18e34970 COW tokens (34345.0) to 0x6C642caFCbd9d8383250bb25F67aE409147f78b2
Redirecting solver 0x7A0A8890D71A4834285efDc1D18bb3828e765C6A COW tokens (360.0) to 0x6C642caFCbd9d8383250bb25F67aE409147f78b2
Redirecting solver 0x8567351D6989d83513D3BC3ad951CcCe363941e3 COW tokens (595.0) to 0x28Dc5cd8785A31199211205124d140B57b054108
Redirecting solver 0xa0044c620Da7f2876dA7004719b8370eb7BE5e50 COW tokens (850.0) to 0x6C642caFCbd9d8383250bb25F67aE409147f78b2
Redirecting solver 0xb20B86C4e6DEEB432A22D773a221898bBBD03036 COW tokens (35040.0) to 0x6C642caFCbd9d8383250bb25F67aE409147f78b2
Redirecting solver 0xc9ec550BEA1C64D779124b23A26292cc223327b6 COW tokens (112875.0) to 0x84dbAE2549d67CaF00f65C355de3D6F4df59A32C
Redirecting solver 0xda324c2f06d3544E7965767ce9ca536dcb67a660 COW tokens (1455.0) to 0x6C642caFCbd9d8383250bb25F67aE409147f78b2
Redirecting solver 0xdA869Be4AdEA17AD39E1DFeCe1bC92c02491504f COW tokens (21375.0) to 0x6C642caFCbd9d8383250bb25F67aE409147f78b2
Redirecting solver 0xdAE69AffE582d36f330Ee1145995A53Fab670962 COW tokens (460.0) to 0x6C642caFCbd9d8383250bb25F67aE409147f78b2
Redirecting solver 0xDe786877a10DBb7EBa25a4DA65aEcf47654F08ab COW tokens (85.0) to 0x6C642caFCbd9d8383250bb25F67aE409147f78b2
Redirecting solver 0xe18B5632DF2Ec339228DD65e4D9F004eF59653d3 COW tokens (10480.0) to 0x28Dc5cd8785A31199211205124d140B57b054108
Deducting 224512758348341870592 COW from reward for 0xe33062a24149f7801a48B2675Ed5111d3278F0f5
Redirecting solver 0xe33062a24149f7801a48B2675Ed5111d3278F0f5 COW tokens (115.48724165165812) to 0x6C642caFCbd9d8383250bb25F67aE409147f78b2
Redirecting solver 0xE9aE2D792F981C53EA7f6493A17Abf5B2a45a86b COW tokens (18925.0) to 0x6C642caFCbd9d8383250bb25F67aE409147f78b2
Redirecting solver 0xF7995B6B051166eA52218c37b8d03A2A6bbeF3DA COW tokens (4320.0) to 0x6C642caFCbd9d8383250bb25F67aE409147f78b2
Total Slippage deducted (ETH): -7.699048214298761
Total ETH Funds needed: 38.71062626688295
Total COW Funds needed: 344355.48724165163
```

</details>

This was accomplished by introducing a `LogSaver` which stores a list of all "print" statements and has the ability to return them at will on demand (which is done for the slack message).

Additionally a few constants were updated.

# Test Plan:

An e2e test was introduced (that should not be run regularly - It actually posts messages to slack!)

You can see the results [here](https://cowservices.slack.com/archives/C03PW4CR38A/p1658934084927009) (It would be possible to delete these messages after the test if we want to go that way). 

Screenshot 2022-07-27 at 10 46 27 PM

[Here](https://cowservices.slack.com/archives/C03PW4CR38A/p1658953802134659) is a real test run posting to Rinkeby.

Screenshot 2022-07-27 at 10 45 56 PM

along with the corresponding [proposed transactions](https://gnosis-safe.io/app/rin:0x8990c564ec303C7b26d3d5556ef0910E58Be08Ce/transactions/queue). Notice that running this script twice results in overwriting the unconfirmed transaction.

Screenshot 2022-07-27 at 10 45 27 PM

### CODE REVIEW

File: src/constants.py [hunk 1]
```diff
@@ -8,10 +8,13 @@
 from web3 import Web3
```
Comment (gentrexha) [inline src/constants.py:8]: What's the package for `gnosis` in `requirements.txt`?

Comment (gentrexha) [reply src/constants.py:8]: Ah, I see it's installed as a dependency from another package. Do we have to define it in the `requirements.txt` then?

File: src/fetch/transfer_file.py [hunk 2]
```diff
@@ -389,35 +398,69 @@ def unusual_slippage_url(period: AccountingPeriod) -> str:
...
+    # Post logs in thread.
+    slack_client.chat_postMessage(
+        channel=channel,
+        format="mrkdwn",
+        text=sub_message,
+        # According to https://api.slack.com/methods/conversations.replies
+        thread_ts=response.get("ts"),
+        unfurl_media=False,
+    )
+    # return response.get("ts"), sub_response.get("ts")
```
Comment (gentrexha) [inline src/fetch/transfer_file.py]: Forgotten comment?

**Review by gentrexha — APPROVED** [review]
Except that one possible forgotten comment, LGTM! Nice work!

File: src/constants.py (hunk 1 above)
Comment (bh2smith) [reply src/constants.py:8]: You may have highlighted the wrong line here because this is a web3 import.

The gnosis library that your IDE is insisting you install is actually an unknown library (that is not supported by gnosis and is suspected to be spam). The gnosis package is installed and included with `safe-eth-py`. 

I asked the same question to @Uxio0 when putting this together.

File: src/fetch/transfer_file.py (hunk 2 above)
Comment (bh2smith) [reply src/fetch/transfer_file.py]: Kinda. I am actually still considering Writing a cleanup option in our unit tests but do do this, the post_to_slack would have to return a message ID that can be deleted after. Unfortunately these values are not deterministic, so I would also have to come up with a better way to test that the message was sent (instead of assert = None <-- the current return value of the method). I have been playing with the idea of returning the IDs of the successfully posted messages so that we can refer back to them. Still not sure yet what to do about this. I might actually leave this comment and add a TODO with some more context here for a follow up PR.

Comment (bh2smith) [reply src/fetch/transfer_file.py]: Created an issue for this: https://github.com/cowprotocol/solver-rewards/issues/81

**Review by josojo — APPROVED** [review]
Cool
